accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -0,0 +1,1212 @@
+
+An instance creation expression of the form
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;(...)</code>
+or
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;.name(...)</code>
+is used to invoke these constructors, and the type of such an expression is
+<code>Name&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>.
+
+*In short, extension type constructors appear to be very similar to
+constructors in classes, and they correspond to the situation where the
+enclosing class has a single, non-late, final instance variable, which is
+initialized according to the normal rules for constructors (in particular,
+it can occur by means of `this.id`, or in an initializer list, but it is an
+error if the initialization does not occur at all).*
+
+An extension type `V` used as an expression (*a type literal*) is allowed
+and has static type `Type`.
+
+*Class modifiers can not be used with extension types. As the grammar
+shows, any occurrence of the keywords `abstract`, `final`, `base`,
+`interface`, `sealed`, or `mixin` in an extension type declaration header
+is a syntax error.*
+
+
+### Composing Extension Types
+
+This section describes the effect of including a clause derived from
+`<interfaces>` in an extension type declaration. We use the phrase
+_the implements clause_ to refer to this clause.
+
+*The rationale is that the set of members and member implementations of a
+given extension type may need to overlap with that of other extension type
+declarations. The implements clause allows for implementation reuse by
+putting some shared members in an extension type `V`, and including `V` in
+the implements clause of several extension type declarations
+<code>V<sub>1</sub> .. V<sub>k</sub></code>, thus "inheriting" the members
+of `V` into all of <code>V<sub>1</sub> .. V<sub>k</sub></code> without code
+duplication.*
+
+*The reason why this mechanism uses the keyword `implements` rather
+than `extends` to declare a relation that involves "inheritance" is
+that it has a similar semantics as that of extension members (in that
+they are statically resolved, and each member is applicable to every
+subtype).*
+
+Assume that _DV_ is an extension type declaration named `Name`, and
+`V1` occurs as one of the `<type>`s in the `<interfaces>` of _DV_. In
+this case we say that `V1` is a _superinterface_ of _DV_.
+
+If _DV_ does not include an `<interfaces>` clause then _DV_ has 
+`Object?` or `Object` as a direct superinterface, according to the subtype
+relation which was specified earlier.
+
+A compile-time error occurs if `V1` is a type name or a parameterized type
+which occurs as a superinterface in an extension type declaration _DV_, but
+`V1` does not denote an extension type, and `V1` does not denote a
+supertype of the ultimate representation type of _DV_.
+
+A compile-time error occurs if any direct or indirect superinterface
+of _DV_ is the type `Name` or a type of the form `Name<...>`. *As
+usual, subtype cycles are not allowed.*
```

Comment:
How about:
```dart
extension type A<T>(Object? _) {}
extension type C<T>(List<Never> _) implements A<List<C<T>>> {}
```
The last line has an infinite expansion, but only in the superinterface hieararchy.

(I have no idea if it works or not.)

Reply:
I don't think that's a problem, we do not ever iterate over all supertypes, we only traverse superinterface graph paths, and they only have an edge when we have a syntactic relation expressed using a clause like `extends`, `implements`, `on`, or `with`.

One reason why it does not seem to be dangerous is that we can have this, no problems:

```dart
class A<T> {
  final Object? rep;
  A(this.rep);
}

class C<T> implements A<List<C<T>>> {
  final List<Never> rep;
  C(this.rep);
}
```